### ISSUE TITLE
Enable source side transformation

### ISSUE THREAD

**Issue by igarashitm** [atlasmap/atlasmap#278]
Today we only have target side transformation (field action) enabled in UI. But some of transformations (e.g. NumberFieldActions) expects a collection as an input and produce non-collection output. This is not possible with target transformation as it only applies with target field type.

Source:(Array[Number]) -> SumUp transformation(Array[Number]->Number) -> Target:(Number)

To utilize those transformations, we need to enable source side transformation. Apparently we already have it in runtime, hopefully it just works once it's enabled in UI.

**Comment by jpav** [comment]
I would suggest that maybe we need **_only_** source-side transformations, with the last transformation being a map/split/combine transformation (probably involving the grouping transformation that would implemented by #244).

**Comment by igarashitm** [comment]
Look&feel / wording aside, we do need both of pre-mapping and post-mapping transformation, e.g. String->Date mapping with `AddDays` transformation.

**Comment by jpav** [comment]
I disagree.  That example can be accomplished just as easily by doing 1) String->Date conversion, 2) Add days to result, and finally 3) map result to target.

**Comment by mattrpav** [comment]
I agree with that supporting FieldActions on both source and target fields is a useful feature

**Comment by jpav** [comment]
Is there an example you can think of where it would be better with both? It seems to me that having both complicates the user experience, and most likely the code base.

**Comment by mattrpav** [comment]
@jpav not sure how it would complicate the code base. Seems to me it should use same code path as target field action processing?

I believe it’s common in other mapping engines and if you take the transform, then split or map to multiple output it would be handy to perform the action once

**Comment by jpav** [comment]
The main thing I was thinking of with the code base is having to constantly be aware of which 'field' context you're dealing with, source or target, and potentially having different rules for each. But maybe that's an over concern. Not sure if I'd say other mapping engines doing this is a good argument. Almost every other one I've seen sucks :-) For what it's worth, the one's I've worked on with Teiid and ModeShape didn't work that way. But considering UX, I'm mainly thinking it would be an awfully simple and straight forward approach to have the workflow be choose source(s), optionally perform one or more transformations on the data, and lastly map to the appropriate targets. I'm also considering possible complex scenarios where there may be multiple chained transformations involved, along with multiple transformation paths from different sources before the final values are set in the target fields. Allowing the mappings to appear anywhere in the middle of this process just seems to complicate things.

**Comment by jpav** [comment]
Also keep in mind transformations performed after the target mapping would have to have the same type/collection type for the target field, transformation source, and transformation target. Ran into this today with #368.

**Comment by mattrpav** [comment]
This feels like two questions:

1) Should the engine support field actions on both source and target? 

I think this one is easier and supporting it does not necessarily have any impact on the UI/UX.  Given, the scenarios where it makes sense may be of the more complex variety.. perhaps many of those mappings are done by hand? Other products supporting it doesn't mean that they have it correct, I read that as they support it from a "we want that capability" or "we've run across that use case". Other products "sucking" doesn't invalidate the need for the capability (imho) and not supporting it seems to be introducing an arbitrary limitation of the backend framework.

2) Should the UI support it? 

I think this is where the nuance comes in. I agree, 80/20 applies here and 80% of use cases are straight-forward and it is not desirable to have a cluttered UI in order to support all-potential use cases. I'm not involved in the same UI/UX and end-user feedback discussions so my opinion as it relates to Syndesis is not as informed. My general opinion is that it could be supported and not get in the way by hiding the second field action panel behind an 'advanced' menu or similar to keep the clean UI look for the majority of use cases.

TLDR; I agree with @igarashitm

**Comment by igarashitm** [comment]
I know having both of pre-mapping and post-mapping introduce a bit of complexity from UX POV, but #277 would be a good idea to solve it, visualize a tiny workflow in the canvas.

 > I disagree. That example can be accomplished just as easily by doing 1) String->Date conversion, 2) Add days to result, and finally 3) map result to target.

1) String->Date conversion in this flow is yet another complexity and I don't think it's worth introducing. And no idea how it's supposed to be represented in UI.

**Comment by jpav** [comment]
@mattrpav My comment about "other products suck" was just tongue and cheek, not a real argument. I'm more coming from my experience with working on transformation engines. I guess bottom line is I'd love to see a use-case from the 20% to justify the need for post-mapping transformations rather than assume they might be needed.

**Comment by jpav** [comment]
@igarashitm 

> String->Date conversion in this flow is yet another complexity and I don't think it's worth introducing. And no idea how it's supposed to be represented in UI.

Can you expound on this? Seems like a straight-forward auto-conversion to me using the 'parse' method from the applicable date type.

**Comment by igarashitm** [comment]
From where do you invoke it? runtime is not built like that. Are you saying you want to rewrite AtlasMap from scratch?

**Comment by pleacu** [comment]
As the original issue with this is approaching tl;dr critical mass - can we restate where we're at?

1.  We clearly need source-side transformations.  
2.  I believe we clearly need target side post-mapping transformations.  
3.  We will be shipping target side transformations so they're in the user domain - they're here to stay.
4.  We could potentially use the 'Show/ hide mapping table' table to let the user see a graphical representation of any one specific mapping.
5. If a complicated UI is an issue, cleaning up the 'Index' references for combine and separate will help unclutter the UI a lot.

wdyt?

**Comment by jpav** [comment]
After a good discussion with @igarashitm, I'm going to concede the need for post-mapping transformations, mainly to allow for auto-conversion during the mapping step. After thinking about it more, my main concern is really how all of these future enhancements end up looking to the user. But that's a separate issue to discuss in the future...

**Comment by jpav** [comment]
Fixed by issue #507
